### PR TITLE
[train] Add TorchAwsNeuronXLABackend and XLAConfig

### LINKED ISSUE #39129: [Train] Add TorchXLA config and backend
### Description

Currently, Train doesn't support XLA config/backend for Torch. AWS Neuron core uses [XLA](https://github.com/pytorch/xla) and with this addition we would add new Torch XLAConfig and appropriate setup to run torch on Tranium instances. 

### Use case

This would benefit users to avoid custom configuration for training on Tranium instances.

### PR DESCRIPTION
**Pull request by chappidim**
## Why are these changes needed?
This change adds new TorchXLA config and NeuronBackend with XLA setup. We initialize the setup with appropriate environment variables and the distributed group.
Please note, torch-neuronx (aws provided) doesn't support pjrt and currently uses xrt server.

## Related issue number

* "#33504"
* "Closes #39129"

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Manual testing

### Manual testing
* Single trn1.2xl machine with 2 neuron_cores
* Train function: All_reduce 
```
from ray.air import session
import ray
from ray.train.backend import Backend, BackendConfig
from ray.train._internal.utils import get_address_and_port
from ray.train.torch.config import TorchXLAConfig
from ray.train._internal.worker_group import WorkerGroup
from ray.train.torch import TorchTrainer

#
import os
import torch
import uuid

ray.init()


class NN(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.layer1 = torch.nn.Linear(4, 4)
        self.nl1 = torch.nn.ReLU()
        self.layer2 = torch.nn.Linear(4, 2)
        self.nl2 = torch.nn.Tanh()

    def forward(self, x):
        x = self.nl1(self.layer1(x))
        return self.nl2(self.layer2(x))


def log(txt):
    rank = os.environ.get("RANK", "unk")
    print(f"{rank}: {txt}", flush=True)


def train_func():
    import torch_xla.core.xla_model as xm

    log("before 1st rendezvous")
    xm.rendezvous('first')
    device = xm.xla_device()
    for c in range(1000):
        ones = torch.ones((2, 3))
        xones = ones.to(device)
        result = xm.all_reduce('sum', xones)
        xm.mark_step()
        result_cpu = result.cpu()
        expected = torch.ones((2, 3)) * int(os.environ.get("WORLD_SIZE", 0))
        log(f"result: {c}: {result}  result.size(): {result.size()}")
        assert torch.all(result_cpu == expected), f'ERROR: {result_cpu} != {expected}'
    log("before final rendezvous")
    xm.rendezvous('last')
    log("done!")


train_dataset = ray.data.from_items([1, 2, 3])
assert train_dataset.count() == 3
trainer = TorchTrainer(
    train_loop_per_worker=train_func,
    torch_config=TorchXLAConfig(),
    scaling_config=ray.air.config.ScalingConfig(num_workers=2, resources_per_worker={"neuron_cores": 1}),
    datasets={"train": train_dataset},
)
result = trainer.fit()
print(result)

```
* Train results
```

 Memory Usage Summary
 Host Used Memory          Total: 426.0MB                    Tensors: 0.0B                     Constants: 0.0B                   DMA Buffers: 128.0KB              App. Memory: 425.9MB
 Device Used Memory        Total: 1.5GB                      Tensors: 120.0B                   Constants: 334.0MB                Model Code: 192.3MB               Runtime Memory: 2.1KB             Model Scratchpad: 1.0GB

 Memory Usage Details
                                                                                                                                Model ID                          Device Memory                     Host Memory
  [-] ND 0                                                                                                                                                         1.5GB                             40.0KB
      [+] NC 0                                                                                                                                                     775.2MB                           20.0KB
      [+] NC 1                                                                                                                                                     775.2MB                           20.0KB


```

### CODE REVIEW

File: python/ray/train/torch/config.py
Comment (matthewdeng) [inline python/ray/train/torch/config.py:60]: What are the supported values here?

File: python/ray/train/torch/config.py
Comment (matthewdeng) [inline python/ray/train/torch/config.py:242]: Will this be supported in this PR? Should we instead only restrict everything to Neuron for now to keep it simple and concise, and extend support for this in the future?

File: python/ray/train/torch/config.py
Comment (matthewdeng) [inline python/ray/train/torch/config.py:265]: `ray.air.session` was recently renamed. See updated `TorchBackend`:

https://github.com/ray-project/ray/blob/85186b67b4c45ec8ecb7344e63e28476baf81531/python/ray/train/torch/config.py#L130-L146

File: python/ray/train/torch/config.py
Comment (chappidim) [reply python/ray/train/torch/config.py]: Added for future use-case and I was expecting some generic setup for TPU and NeuronCores as they both use TorchXLA to some extent. Removed for now.

File: python/ray/train/torch/config.py
Comment (chappidim) [reply python/ray/train/torch/config.py]: 1. Added, it's xrt/pjrt.
2. Also, refactor to accelerator_type and included neuron_cores

File: python/ray/train/torch/xla/config.py
Comment (matthewdeng) [inline python/ray/train/torch/xla/config.py:16]: I think this should subclass `TorchConfig`?

File: python/ray/train/torch/xla/config.py
Comment (matthewdeng) [inline python/ray/train/torch/xla/config.py:123]: This logic should be run after training.

File: python/ray/train/torch/xla/config.py
Comment (matthewdeng) [inline python/ray/train/torch/xla/config.py:75]: Can you put the specific numbers, e.g. `# noqa: F401`

File: python/ray/train/torch/xla/config.py
Comment (matthewdeng) [inline python/ray/train/torch/xla/config.py:83]: Should this subclass `TorchBackend`?

File: python/ray/train/torch/xla/config.py
Comment (chappidim) [reply python/ray/train/torch/xla/config.py:57]: I haven't seen any better documentation to be honest and my knowledge is slim in this area.

Also, do we have a way to pass addition work/actions before the training starts without extending the class? For example, I want to add a new EFA configuration as env_var for this train/run. More like a function to `on_start` which will be executed on all workers.

File: python/ray/train/torch/xla/config.py
Comment (chappidim) [reply python/ray/train/torch/xla/config.py:84]: It's private class, kept to Backend for now

File: python/ray/train/torch/xla/config.py
Comment (amogkam) [inline python/ray/train/torch/xla/config.py:38]: if other accelerator types are not supported, do we need to make `accelerator_type` configurable at all?

File: python/ray/train/torch/xla/config.py
Comment (amogkam) [inline python/ray/train/torch/xla/config.py:44]: is there a python API we can use to kill the server?

File: python/ray/train/torch/xla/config.py
Comment (amogkam) [inline python/ray/train/torch/xla/config.py:93]: what other configurations are possible other than xrt? if only xrt is supported do we need to make it configurable?

File: python/ray/train/torch/xla/config.py
Comment (chappidim) [reply python/ray/train/torch/xla/config.py:38]: It can extendible for TPU (similar class of xla).

File: python/ray/train/torch/xla/config.py
Comment (chappidim) [reply python/ray/train/torch/xla/config.py:44]: Unfortunately, no. I'm open for options if there are any alternatives.

File: python/ray/train/torch/xla/config.py
Comment (chappidim) [reply python/ray/train/torch/xla/config.py:93]: Listed in docs (xrt, pjrt). pjrt is not yet supported for NeuronCore and is WiP.

**Comment by chappidim** [comment]
> @chappidim One concern is that these features have not yet been tested. Is it possible for you to provide AWS Neuron instances and test these code? (Our current CI is running on general `g4dn` and `m5` instances.)

I did manual testing for all_reduce function mentioned in PR overview. I'm happy to run a training script of your choice on trn1.32xl instance/s (limit=2) if needed. Let me know.

**Comment by matthewdeng** [comment]
cc @can-anyscale would it be possible for us to add `trn1.32xl` instances to use in our CI tests?

**Comment by matthewdeng** [comment]
BTW we are planning to add `trn1.2xlarge` instances to our CI pipeline next week to be able to test this cc @can-anyscale

**Comment by can-anyscale** [comment]
@aslonnie is helping me with this

File: python/ray/train/torch/xla/config.py
Comment (matthewdeng) [reply python/ray/train/torch/xla/config.py:57]: Okay can we just add a TODO here. If we need to support heterogenous cluster we can add more robust logic in `on_training_start` to fetch the topology of the cluster and then set it.

For the question on `on_start`, just capturing here the conversation we had offline - this can be defined by the user in in `train_loop_per_worker`.

File: python/ray/train/torch/xla/config.py
Comment (matthewdeng) [reply python/ray/train/torch/xla/config.py:37]: Can we just `runtime` and `accelerator_type` attributes in this PR to keep the surface area simple? We can document in the class docstring that only this combination is supported right now.

Once we want to add  support TPU and/or pjrt it will be better to make sure the interface and testing is comprehensive then.

Comment (chappidim) [reply python/ray/train/torch/xla/config.py:37]: @matthewdeng To confirm, we want to remove runtime/acclerator_type and move the information to docs?

Comment (matthewdeng) [reply python/ray/train/torch/xla/config.py:37]: Yeah more specifically can just be in the `TorchXLAConfig` docstring - I think this would be the simplest and cleanest solution fro now.

**Comment by cosnicolaou** [comment]
Any progress on this?

**Comment by woshiyyya** [comment]
@cosnicolaou we are working on make multi-node training works. Will merge this PR after the release test is passed.

**Review by woshiyyya — APPROVED** [review]

File: release/ray_release/byod/byod_train_trainium.sh
Comment (can-anyscale) [inline release/ray_release/byod/byod_train_trainium.sh:33]: we need to add this to https://github.com/ray-project/ray/blob/master/release/ray_release/byod/requirements_byod_3.9.in and recompile it so we can have all transition dependencies pinned

Comment (anyscalesam) [reply release/ray_release/byod/byod_train_trainium.sh:33]: @can-anyscale was this ever done?

Comment (can-anyscale) [reply release/ray_release/byod/byod_train_trainium.sh:33]: not yet, need to be done this in PR

Comment (can-anyscale) [reply release/ray_release/byod/byod_train_trainium.sh:33]: i tried to compile the dependencies together but neuronx-cc requires `networkx<=2.6.3` which conflicts with ray docker image; so these packages need to be installed in its byod file which is what we already doing here

let's merge with this

File: release/train_tests/trainium/compute_aws.yaml
Comment (aslonnie) [inline release/train_tests/trainium/compute_aws.yaml:1]: what is this cloud? should we add some doc here on how this cloud is provisioned? or where does it come from?

like if some of the subnets here are broken or something, what should one do?

Comment (woshiyyya) [reply release/train_tests/trainium/compute_aws.yaml:1]: This is the `test-efa-vpn-cloud`, we need EFA setup to run Neuron code.

Comment (aslonnie) [reply release/train_tests/trainium/compute_aws.yaml:1]: can you add the related info in the form of yaml comments?

like how to create / maintain this cloud and its related networks, and where to find it.

Comment (woshiyyya) [reply release/train_tests/trainium/compute_aws.yaml:1]: Ah we are going to delete this release test, as we've decided to make AWSNeuron Integration as community supported.

File: python/ray/train/torch/xla/config.py
Comment (woshiyyya) [inline python/ray/train/torch/xla/config.py:169]: Why are we compiling the graph at the end? How would the people use the compiled graph?

**Review by woshiyyya — CHANGES_REQUESTED** [review]
The CI tests are failing. Can you fix it before we merge it?

File: python/ray/train/torch/xla/config.py
Comment (5cp) [reply python/ray/train/torch/xla/config.py:169]: This code only runs during a pre-compilation step which happens when neuron_parallel_compile is set to true. `_neuron_compile_extracted_graphs()` must run at the end of the job (after a short number of training iterations) when all the graphs have been encountered.

See: [1](https://github.com/5cp/data-on-eks/tree/kuberay_trn1/ai-ml/trainium-inferentia/examples/kuberay_llama2_pretrain#run-precompilation-job) and [2](https://github.com/5cp/data-on-eks/blob/kuberay_trn1/ai-ml/trainium-inferentia/examples/kuberay_llama2_pretrain/docker/llama2_pretrain/launch_precompile_job.sh#L3) and [3](https://github.com/5cp/data-on-eks/blob/kuberay_trn1/ai-ml/trainium-inferentia/examples/kuberay_llama2_pretrain/docker/llama2_pretrain/ray_train_llama2.py#L198).

After precompilation, the user simply runs without neuron_parallel_compile to use the cached graphs and this avoids `_neuron_compile_extracted_graphs()`

File: release/train_tests/trainium/compute_aws.yaml
Comment (anyscalesam) [reply release/train_tests/trainium/compute_aws.yaml]: Can we resolve this then @woshiyyya @aslonnie ?

File: release/ray_release/byod/byod_train_trainium.sh
Comment (anyscalesam) [reply release/ray_release/byod/byod_train_trainium.sh]: Superb; can we resolve this Comment thread @can-anyscale ?

File: python/ray/train/torch/xla/config.py
Comment (anyscalesam) [reply python/ray/train/torch/xla/config.py:169]: @woshiyyya does the above make sense?

**Comment by woshiyyya** [comment]
There are two Train premerge tests [[1](https://buildkite.com/ray-project/premerge/builds/22370#018e5d94-bbeb-4500-bb17-1f810f2f406b), [2](https://buildkite.com/ray-project/premerge/builds/22370#018e5d94-bc15-4668-aa92-ec5f08ba0b2f)] are failing. 

Screenshot 2024-03-25 at 12 22 01 PM

**Review by woshiyyya — APPROVED** [review]

File: release/train_tests/trainium/compute_aws.yaml
Comment (woshiyyya) [reply release/train_tests/trainium/compute_aws.yaml]: Yes. The test is removed.

**Pull request merged**
